File: EngineQ/EngineQCommon/Math/Matrix3.cpp
#include "Matrix3.hpp"

#include "Vector3.hpp"

namespace EngineQ
{
	namespace Math
	{
	#pragma region Constructors

		Matrix3::Matrix3(Real value) :
			M00{ value }, M01{ value }, M02{ value },
			M10{ value }, M11{ value }, M12{ value },
			M20{ value }, M21{ value }, M22{ value }
		{
		}

		Matrix3::Matrix3(
			Real m00, Real m01, Real m02,
			Real m10, Real m11, Real m12,
			Real m20, Real m21, Real m22) :
			M00{ m00 }, M01{ m01 }, M02{ m02 },
			M10{ m10 }, M11{ m11 }, M12{ m12 },
			M20{ m20 }, M21{ m21 }, M22{ m22 }
		{
		}

		Matrix3::Matrix3(const Vector3& row0, const Vector3& row1, const Vector3& row2) :
			M00{ row0.X }, M01{ row0.Y }, M02{ row0.Z },
			M10{ row1.X }, M11{ row1.Y }, M12{ row1.Z },
			M20{ row2.X }, M21{ row2.Y }, M22{ row2.Z }
		{
		}

	#pragma endregion

	#pragma region Methods

		void Matrix3::Transpose()
		{
			Real tmp;
			tmp = M01; M01 = M10; M10 = tmp;
			tmp = M02; M02 = M20; M20 = tmp;
			tmp = M12; M12 = M21; M21 = tmp;
		}

		void Matrix3::Inverse()
		{
			*this = GetInversed();
		}

		std::string Matrix3::ToString() const
		{
			return Utilities::ToString(*this);
		}
		
		Matrix3 Matrix3::GetTransposed() const
		{
			return Matrix3{ M00, M10, M20, M01, M11, M21, M02, M12, M22 };
		}

		Matrix3 Matrix3::GetInversed() const
		{
			Real tmp1 = M11 * M22 - M12 * M21;
			Real tmp2 = M12 * M20 - M10 * M22;
			Real tmp3 = M10 * M21 - M11 * M20;

			Real det = M00 * tmp1 + M01 * tmp2 + M02 * tmp3;
			
			if (det == static_cast<Real>(0))
				return Matrix3{};
			
			Real invdet = static_cast<Real>(1) / det;

			return Matrix3{
				tmp1 * invdet, (M02 * M21 - M01 * M22) * invdet, (M01 * M12 - M02 * M11) * invdet,
				tmp2 * invdet, (M00 * M22 - M02 * M20) * invdet, (M02 * M10 - M00 * M12) * invdet,
				tmp3 * invdet, (M01 * M20 - M00 * M21) * invdet, (M00 * M11 - M01 * M10) * invdet };
		}

		Real Matrix3::GetDeterminant() const
		{
			return (M00 * (M11 * M22 - M12 * M21) + M01 * (M12 * M20 - M10 * M22) + M02 * (M10 * M21 - M11 * M20));
		}

		Vector3 Matrix3::GetDiagonal() const
		{
			return Vector3{ M00, M11, M22 };
		}

		void Matrix3::SetDiagonal(const Vector3& diagonal)
		{
			M00 = diagonal.X;
			M11 = diagonal.Y;
			M22 = diagonal.Z;
		}

		Vector3 Matrix3::GetColumn(std::size_t column) const
		{
			return Vector3{ Values2[0][column], Values2[1][column], Values2[2][column] };
		}

		void Matrix3::SetColumn(std::size_t column, Vector3 value)
		{
			Values2[0][column] = value.X;
			Values2[1][column] = value.Y;
			Values2[2][column] = value.Z;
		}

		Vector3 Matrix3::GetRow(std::size_t row) const
		{
			return Vector3{ Values2[row][0], Values2[row][1], Values2[row][2] };
		}

		void Matrix3::SetRow(std::size_t row, Vector3 value)
		{
			Values2[row][0] = value.X;
			Values2[row][1] = value.Y;
			Values2[row][2] = value.Z;
		}

		void Matrix3::GetRows(Vector3& row0, Vector3& row1, Vector3& row2) const
		{
			row0.X = M00; row0.Y = M01; row0.Z = M02;
			row1.X = M10; row1.Y = M11; row1.Z = M12;
			row2.X = M20; row2.Y = M21; row2.Z = M22;
		}

		void Matrix3::GetColumns(Vector3& column0, Vector3& column1, Vector3& column2) const
		{
			column0.X = M00; column0.Y = M10; column0.Z = M20;
			column1.X = M01; column1.Y = M11; column1.Z = M21;
			column2.X = M02; column2.Y = M12; column2.Z = M22;
		}

	#pragma endregion

	#pragma region Static Methods

		Matrix3 Matrix3::GetIdentity()
		{
			return Matrix3{
				static_cast<Real>(1), static_cast<Real>(0), static_cast<Real>(0),
				static_cast<Real>(0), static_cast<Real>(1), static_cast<Real>(0),
				static_cast<Real>(0), static_cast<Real>(0), static_cast<Real>(1) };
		}

		Matrix3 Matrix3::CreateFromColumns(const Vector3& column0, const Vector3& column1, const Vector3& column2)
		{
			return Matrix3{
				column0.X, column1.X, column2.X,
				column0.Y, column1.Y, column2.Y,
				column0.Z, column1.Z, column2.Z 
			};
		}

		Matrix3 Matrix3::CreateFromRows(const Vector3& row0, const Vector3& row1, const Vector3& row2)
		{
			return Matrix3{
				row0.X, row0.Y, row0.Z,
				row1.X, row1.Y, row1.Z,
				row2.X, row2.Y, row2.Z 
			};
		}

	#pragma endregion

	#pragma region Operators

		void Matrix3::operator *= (const Matrix3& matrix)
		{
			*this = *this * matrix;
		}

		void Matrix3::operator *= (Real scalar)
		{
			M00 *= scalar;
			M01 *= scalar;
			M02 *= scalar;
			M10 *= scalar;
			M11 *= scalar;
			M12 *= scalar;
			M20 *= scalar;
			M21 *= scalar;
			M22 *= scalar;
		}

		void Matrix3::operator /= (Real scalar)
		{
			*this *= static_cast<Real>(1) / scalar;
		}

		Real Matrix3::operator () (std::size_t index) const
		{
			return Values[index];
		}

		Real& Matrix3::operator () (std::size_t index)
		{
			return Values[index];
		}

		Real Matrix3::operator [] (std::size_t index) const
		{
			return Values[index];
		}

		Real& Matrix3::operator [] (std::size_t index)
		{
			return Values[index];
		}

		Real Matrix3::operator () (std::size_t row, std::size_t column) const
		{
			return Values2[row][column];
		}

		Real& Matrix3::operator () (std::size_t row, std::size_t column)
		{
			return Values2[row][column];
		}

	#pragma endregion

	#pragma region Static Operators

		Matrix3 operator * (const Matrix3& lhs, const Matrix3& rhs)
		{
			return Matrix3{
				lhs.M00 * rhs.M00 + lhs.M01 * rhs.M10 + lhs.M02 * rhs.M20,
				lhs.M00 * rhs.M01 + lhs.M01 * rhs.M11 + lhs.M02 * rhs.M21,
				lhs.M00 * rhs.M02 + lhs.M01 * rhs.M12 + lhs.M02 * rhs.M22,

				lhs.M10 * rhs.M00 + lhs.M11 * rhs.M10 + lhs.M12 * rhs.M20,
				lhs.M10 * rhs.M01 + lhs.M11 * rhs.M11 + lhs.M12 * rhs.M21,
				lhs.M10 * rhs.M02 + lhs.M11 * rhs.M12 + lhs.M12 * rhs.M22,

				lhs.M20 * rhs.M00 + lhs.M21 * rhs.M10 + lhs.M22 * rhs.M20,
				lhs.M20 * rhs.M01 + lhs.M21 * rhs.M11 + lhs.M22 * rhs.M21,
				lhs.M20 * rhs.M02 + lhs.M21 * rhs.M12 + lhs.M22 * rhs.M22
			};
		}

		Vector3 operator * (const Matrix3& matrix, const Vector3& vector)
		{
			return Vector3{
				matrix.M00 * vector.X + matrix.M01 * vector.Y + matrix.M02 * vector.Z,
				matrix.M10 * vector.X + matrix.M11 * vector.Y + matrix.M12 * vector.Z,
				matrix.M20 * vector.X + matrix.M21 * vector.Y + matrix.M22 * vector.Z
			};
		}

		Matrix3 operator * (const Matrix3& matrix, Real scalar)
		{
			Matrix3 ret = matrix;
			ret *= scalar;
			return ret;
		}

		Matrix3 operator * (Real scalar, const Matrix3& matrix)
		{
			Matrix3 ret = matrix;
			ret *= scalar;
			return ret;
		}

		Matrix3 operator / (const Matrix3& matrix, Real scalar)
		{
			Matrix3 ret = matrix;
			ret *= static_cast<Real>(1) / scalar;
			return ret;
		}

		bool operator == (const Matrix3& lhs, const Matrix3& rhs)
		{
		//	return
		//		lhs.M00 == rhs.M00 && lhs.M01 == rhs.M01 && lhs.M02 == rhs.M02 &&
		//		lhs.M10 == rhs.M10 && lhs.M11 == rhs.M11 && lhs.M12 == rhs.M12 &&
		//		lhs.M20 == rhs.M20 && lhs.M21 == rhs.M21 && lhs.M22 == rhs.M22;
			return
				Utils::EpsEqual(lhs.M00, rhs.M00) && Utils::EpsEqual(lhs.M01, rhs.M01) && Utils::EpsEqual(lhs.M02, rhs.M02) &&
				Utils::EpsEqual(lhs.M10, rhs.M10) && Utils::EpsEqual(lhs.M11, rhs.M11) && Utils::EpsEqual(lhs.M12, rhs.M12) &&
				Utils::EpsEqual(lhs.M20, rhs.M20) && Utils::EpsEqual(lhs.M21, rhs.M21) && Utils::EpsEqual(lhs.M22, rhs.M22);
		}

		bool operator != (const Matrix3& lhs, const Matrix3& rhs)
		{
		//	return
		//		lhs.M00 != rhs.M00 || lhs.M01 != rhs.M01 || lhs.M02 != rhs.M02 ||
		//		lhs.M10 != rhs.M10 || lhs.M11 != rhs.M11 || lhs.M12 != rhs.M12 ||
		//		lhs.M20 != rhs.M20 || lhs.M21 != rhs.M21 || lhs.M22 != rhs.M22;
			return
				Utils::EpsNotEqual(lhs.M00, rhs.M00) || Utils::EpsNotEqual(lhs.M01, rhs.M01) || Utils::EpsNotEqual(lhs.M02, rhs.M02) ||
				Utils::EpsNotEqual(lhs.M10, rhs.M10) || Utils::EpsNotEqual(lhs.M11, rhs.M11) || Utils::EpsNotEqual(lhs.M12, rhs.M12) ||
				Utils::EpsNotEqual(lhs.M20, rhs.M20) || Utils::EpsNotEqual(lhs.M21, rhs.M21) || Utils::EpsNotEqual(lhs.M22, rhs.M22);
		}

		std::istream& operator >> (std::istream& stream, Matrix3& matrix)
		{
			stream.ignore(2);
			stream >> matrix.M00;
			stream.ignore(1);
			stream >> matrix.M01;
			stream.ignore(1);
			stream >> matrix.M02;
			stream.ignore(3);
			stream >> matrix.M10;
			stream.ignore(1);
			stream >> matrix.M11;
			stream.ignore(1);
			stream >> matrix.M12;
			stream.ignore(3);
			stream >> matrix.M20;
			stream.ignore(1);
			stream >> matrix.M21;
			stream.ignore(1);
			stream >> matrix.M22;
			stream.ignore(2);

			return stream;
		}

		std::istream& operator >>= (std::istream& stream, Matrix3& matrix)
		{
			return stream.read(reinterpret_cast<char*>(&matrix.Values), sizeof(matrix.Values));
		}

		std::ostream& operator << (std::ostream& stream, const Matrix3& matrix)
		{
			return stream << "[[" <<
				matrix.M00 << "," << matrix.M01 << "," << matrix.M02 << "],[" <<
				matrix.M10 << "," << matrix.M11 << "," << matrix.M12 << "],[" <<
				matrix.M20 << "," << matrix.M21 << "," << matrix.M22 << "]]";
		}

		std::ostream& operator <<= (std::ostream& stream, const Matrix3& matrix)
		{
			return stream.write(reinterpret_cast<const char*>(matrix.Values), sizeof(matrix.Values));
		}

	#pragma endregion
	}
}

File: EngineQ/Source/EngineQScripting/Math/Matrix3.cs
﻿using System;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;

namespace EngineQ.Math
{
	using Real = System.Single;

	[StructLayout(LayoutKind.Sequential)]
	public struct Matrix3
	{
		#region Fields

		public Real M00;
		public Real M01;
		public Real M02;
		public Real M10;
		public Real M11;
		public Real M12;
		public Real M20;
		public Real M21;
		public Real M22;

		#endregion

		#region Constructors

		/// <summary>
		/// Constructs Matrix from individual elements.
		/// </summary>
		/// <param name="m00">Element at row 0, column 0.</param>
		/// <param name="m01">Element at row 0, column 1.</param>
		/// <param name="m02">Element at row 0, column 2.</param>
		/// <param name="m10">Element at row 1, column 0.</param>
		/// <param name="m11">Element at row 1, column 1.</param>
		/// <param name="m12">Element at row 1, column 2.</param>
		/// <param name="m20">Element at row 2, column 0.</param>
		/// <param name="m21">Element at row 2, column 1.</param>
		/// <param name="m22">Element at row 2, column 2.</param>
		public Matrix3(
			Real m00, Real m01, Real m02,
			Real m10, Real m11, Real m12,
			Real m20, Real m21, Real m22)
		{
			this.M00 = m00;
			this.M01 = m01;
			this.M02 = m02;
			this.M10 = m10;
			this.M11 = m11;
			this.M12 = m12;
			this.M20 = m20;
			this.M21 = m21;
			this.M22 = m22;
		}

		/// <summary>
		/// Constructs Matrix from rows.
		/// </summary>
		/// <param name="row0">Vector of elements for row 0.</param>
		/// <param name="row1">Vector of elements for row 1.</param>
		/// <param name="row2">Vector of elements for row 2.</param>
		public Matrix3(Vector3 row0, Vector3 row1, Vector3 row2)
		{
			this.M00 = row0.X;
			this.M01 = row0.Y;
			this.M02 = row0.Z;
			this.M10 = row1.X;
			this.M11 = row1.Y;
			this.M12 = row1.Z;
			this.M20 = row2.X;
			this.M21 = row2.Y;
			this.M22 = row2.Z;

		}

		#endregion

		#region Properties

		/// <summary>
		/// Element at specified row and column.
		/// </summary>
		/// <param name="row">Element row.</param>
		/// <param name="column">Element column.</param>
		/// <returns>Element at specified row and column.</returns>
		public Real this[int row, int column]
		{
			get
			{
				return this[row * 3 + column];
			}
			set
			{
				this[row * 3 + column] = value;
			}
		}

		/// <summary>
		/// Element at specified index.
		/// </summary>
		/// <param name="index">Element index.</param>
		/// <returns>Element at given index.</returns>
		public Real this[int index]
		{
			get
			{
				switch (index)
				{
					case 0:
						return M00;
					case 1:
						return M01;
					case 2:
						return M02;
					case 3:
						return M10;
					case 4:
						return M11;
					case 5:
						return M12;
					case 6:
						return M20;
					case 7:
						return M21;
					case 8:
						return M22;
					default:
						throw new IndexOutOfRangeException("Invalid index");
				}
			}
			set
			{
				switch (index)
				{
					case 0:
						M00 = value;
						break;
					case 1:
						M01 = value;
						break;
					case 2:
						M02 = value;
						break;
					case 3:
						M10 = value;
						break;
					case 4:
						M11 = value;
						break;
					case 5:
						M12 = value;
						break;
					case 6:
						M20 = value;
						break;
					case 7:
						M21 = value;
						break;
					case 8:
						M22 = value;
						break;
					default:
						throw new IndexOutOfRangeException("Invalid index");
				}
			}
		}

		/// <summary>
		/// Creates new <see cref="Matrix3"/> that is transposition of this one.
		/// </summary>
		public Matrix3 Transposed
		{
			get
			{
				return new Matrix3(M00, M10, M20, M01, M11, M21, M02, M12, M22);
			}
		}

		/// <summary>
		/// Creates new <see cref="Matrix3"/> that is inversion of this one.
		/// </summary>
		public Matrix3 Inverse
		{
			get
			{
				Matrix3 value;
				API_Inverse(ref this, out value);
				return value;
			}
		}

		/// <summary>
		/// Computes determinant of this <see cref="Matrix3"/>.
		/// </summary>
		public Real Determinant
		{
			get
			{
				Real value;
				API_Determinant(ref this, out value);
				return value;
			}
		}

		/// <summary>
		/// <see cref="Vector3"/> of values lying on the diagonal of this <see cref="Matrix3"/>.
		/// </summary>
		public Vector3 Diagonal
		{
			get
			{
				return new Vector3(M00, M11, M22);
			}

			set
			{
				M00 = value.X;
				M11 = value.Y;
				M22 = value.Z;
			}
		}

		#endregion

		#region Static Properties
		
		/// <summary>
		/// Identity Matrix.
		/// </summary>
		public static Matrix3 Identity
		{
			get
			{
				return new Matrix3(
					(Real)1, (Real)0, (Real)0,
					(Real)0, (Real)1, (Real)0,
					(Real)0, (Real)0, (Real)1);
			}
		}

		#endregion

		#region Methods

		/// <summary>
		/// Modifies this <see cref="Matrix3"/> by transposition.
		/// </summary>
		public void Transpose()
		{
			API_Transpose(ref this);
		}

		/// <summary>
		/// Modifies this <see cref="Matrix3"/> by inversion.
		/// </summary>
		public void Invert()
		{
			this = Inverse;
		}

		/// <summary>
		/// Returns string representation of this <see cref="Matrix3"/>. 
		/// </summary>
		/// <returns>String representation of Matrix.</returns>
		public override string ToString()
		{
			return $"[[{M00},{M01},{M02}],[{M10},{M11},{M12}],[{M20},{M21},{M22}]]";
		}

		/// <summary>
		/// Determines whether the specified object is equal to the current <see cref="Matrix3"/>.
		/// </summary>
		/// <param name="obj">The object to compare with the current <see cref="Matrix3"/>.</param>
		/// <returns>true if the specified object is equal to the current <see cref="Matrix3"/>; otherwise, false.</returns>
		public override bool Equals(object obj)
		{
			if (!(obj is Matrix3))
				return false;

			return this == (Matrix3)obj;
		}

		/// <summary>
		/// Default hash function.
		/// </summary>
		/// <returns>Default hash value.</returns>
		public override int GetHashCode()
		{
			return base.GetHashCode();
		}

		/// <summary>
		/// <see cref="Vector3"/> representing one of columns of this <see cref="Matrix3"/>.
		/// </summary>
		/// <param name="column">Index of column.</param>
		/// <returns><see cref="Vector3"/> representing column.</returns>
		public Vector3 GetColumn(int column)
		{
			return new Vector3(this[0, column], this[1, column], this[2, column]);
		}

		/// <summary>
		/// Sets content of <paramref name="column"/> with given <see cref="Vector3"/> <paramref name="value"/>.
		/// </summary>
		/// <param name="column">Index of the column to set.</param>
		/// <param name="value"><see cref="Vector3"/> of values to set in the column.</param>
		public void SetColumn(int column, Vector3 value)
		{
			this[0, column] = value.X;
			this[1, column] = value.Y;
			this[2, column] = value.Z;
		}

		/// <summary>
		/// <see cref="Vector3"/> representing one of rows of this <see cref="Matrix3"/>.
		/// </summary>
		/// <param name="row">Index of row.</param>
		/// <returns><see cref="Vector3"/> representing row.</returns>
		public Vector3 GetRow(int row)
		{
			return new Vector3(this[row, 0], this[row, 1], this[row, 2]);
		}

		/// <summary>
		/// Sets content fo <paramref name="row"/>  with given <see cref="Vector3"/> <paramref name="value"/>.
		/// </summary>
		/// <param name="row">Index of the row to set.</param>
		/// <param name="value"><see cref="Vector3"/> of values to set in the row.</param>
		public void SetRow(int row, Vector3 value)
		{
			this[row, 0] = value.X;
			this[row, 1] = value.Y;
			this[row, 2] = value.Z;
		}

		/// <summary>
		/// Converts this <see cref="Matrix3"/> into <see cref="Vector3"/>s of rows.
		/// </summary>
		/// <returns><see cref="Vector3"/> array of rows.</returns>
		public Vector3[] GetRows()
		{
			return new Vector3[3]
			{
				new Vector3(M00, M01, M02),
				new Vector3(M10, M11, M12),
				new Vector3(M20, M21, M22)
			};
		}

		/// <summary>
		/// Converts this <see cref="Matrix3"/> into <see cref="Vector3"/>s of columns.
		/// </summary>
		/// <returns><see cref="Vector3"/> array of columns.</returns>
		public Vector3[] GetColumns()
		{
			return new Vector3[3]
			{
				new Vector3(M00, M10, M20),
				new Vector3(M01, M11, M21),
				new Vector3(M02, M12, M22)
			};
		}

		#endregion

		#region Static Methods
		
		/// <summary>
		/// Creates <see cref="Matrix3"/> from given <see cref="Vector3"/>s representing columns.
		/// </summary>
		/// <param name="column0">First column.</param>
		/// <param name="column1">Second column.</param>
		/// <param name="column2">Third column.</param>
		/// <returns><see cref="Matrix3"/> created from given rows.</returns>
		public static Matrix3 CreateFromColumns(Vector3 column0, Vector3 column1, Vector3 column2)
		{
			return new Matrix3
			(
				column0.X, column1.X, column2.X,
				column0.Y, column1.Y, column2.Y,
				column0.Z, column1.Z, column2.Z
			);
		}

		/// <summary>
		/// Creates <see cref="Matrix3"/> from given <see cref="Vector3"/>s representing rows.
		/// </summary>
		/// <param name="row0">First row.</param>
		/// <param name="row1">Second row.</param>
		/// <param name="row2">Third row.</param>
		/// <returns><see cref="Matrix3"/> created from given rows.</returns>
		public static Matrix3 CreateFromRows(Vector3 row0, Vector3 row1, Vector3 row2)
		{
			return new Matrix3
			(
				row0.X, row0.Y, row0.Z,
				row1.X, row1.Y, row1.Z,
				row2.X, row2.Y, row2.Z
			);
		}

		#endregion

		#region Operators

		public static Matrix3 operator *(Matrix3 matrix1, Matrix3 matrix2)
		{
			Matrix3 value;
			API_MultiplyMatrix(ref matrix1, ref matrix2, out value);
			return value;
		}

		public static Vector3 operator *(Matrix3 matrix, Vector3 vector)
		{
			Vector3 value;
			API_MultiplyVector3(ref matrix, ref vector, out value);
			return value;
		}

		public static Matrix3 operator *(Matrix3 matrix, Real scalar)
		{
			Matrix3 value;
			API_MultiplyScalar(ref matrix, scalar, out value);
			return value;
		}

		public static Matrix3 operator *(Real scalar, Matrix3 matrix)
		{
			Matrix3 value;
			API_MultiplyScalar(ref matrix, scalar, out value);
			return value;
		}

		public static Matrix3 operator /(Matrix3 matrix, Real scalar)
		{
			Matrix3 value;
			API_DivideScalar(ref matrix, scalar, out value);
			return value;
		}

		public static bool operator ==(Matrix3 matrix1, Matrix3 matrix2)
		{
			bool value;
			API_Equal(ref matrix1, ref matrix2, out value);
			return value;
		}

		public static bool operator !=(Matrix3 matrix1, Matrix3 matrix2)
		{
			bool value;
			API_NotEqual(ref matrix1, ref matrix2, out value);
			return value;
		}

		#endregion

		#region API

		[MethodImpl(MethodImplOptions.InternalCall)]
		private static extern void API_Inverse(ref Matrix3 matrix, out Matrix3 value);

		[MethodImpl(MethodImplOptions.InternalCall)]
		private static extern void API_Determinant(ref Matrix3 matrix, out Real value);

		[MethodImpl(MethodImplOptions.InternalCall)]
		private static extern void API_Transpose(ref Matrix3 value);

		[MethodImpl(MethodImplOptions.InternalCall)]
		private static extern void API_MultiplyMatrix(ref Matrix3 lhs, ref Matrix3 rhs, out Matrix3 value);

		[MethodImpl(MethodImplOptions.InternalCall)]
		private static extern void API_MultiplyVector3(ref Matrix3 lhs, ref Vector3 rhs, out Vector3 value);

		[MethodImpl(MethodImplOptions.InternalCall)]
		private static extern void API_MultiplyScalar(ref Matrix3 lhs, Real rhs, out Matrix3 value);

		[MethodImpl(MethodImplOptions.InternalCall)]
		private static extern void API_DivideScalar(ref Matrix3 lhs, Real rhs, out Matrix3 value);

		[MethodImpl(MethodImplOptions.InternalCall)]
		private static extern void API_Equal(ref Matrix3 lhs, ref Matrix3 rhs, out bool value);

		[MethodImpl(MethodImplOptions.InternalCall)]
		private static extern void API_NotEqual(ref Matrix3 lhs, ref Matrix3 rhs, out bool value);

		#endregion
	}
}
